File: file_manager.py
import os
import math

class FileManager:
    """Handles piece-level file I/O for a peer."""

    def __init__(self, peer_id, common_cfg):
        self.peer_id = peer_id
        self.file_name = common_cfg['FileName']
        self.file_size = common_cfg['FileSize']
        self.piece_size = common_cfg['PieceSize']
        self.num_pieces = math.ceil(self.file_size / self.piece_size)
        self.last_piece_size = self.file_size % self.piece_size or self.piece_size

        # Each peer uses its own subdirectory
        self.dir = f"peer_{peer_id}"
        os.makedirs(self.dir, exist_ok=True)
        self.file_path = os.path.join(self.dir, self.file_name)

    def init_file(self, has_file):
        """
        If the peer already has the file, it should be in the subdirectory.
        If not, create an empty placeholder file of the right size.
        """
        if has_file:
            if not os.path.exists(self.file_path):
                print(f"[FileManager] WARNING: peer_{self.peer_id} marked has_file=1 but file not found at {self.file_path}")
        else:
            # Create empty file if it doesn't exist
            if not os.path.exists(self.file_path):
                with open(self.file_path, 'wb') as f:
                    f.write(b'\x00' * self.file_size)

    def get_piece(self, index):
        """Read and return the bytes for a given piece index."""
        offset = index * self.piece_size
        size = self.piece_size if index < self.num_pieces - 1 else self.last_piece_size
        try:
            with open(self.file_path, 'rb') as f:
                f.seek(offset)
                return f.read(size)
        except Exception as e:
            print(f"[FileManager] Error reading piece {index}: {e}")
            return None

    def save_piece(self, index, data):
        """Write piece data to the correct offset in the file."""
        offset = index * self.piece_size
        try:
            with open(self.file_path, 'r+b') as f:
                f.seek(offset)
                f.write(data)
        except Exception as e:
            print(f"[FileManager] Error saving piece {index}: {e}")

File: FileManager.java
import java.io.*;

public class FileManager {
    private int peerId;
    private CommonConfig config;
    private byte[][] pieces;
    private boolean[] hasPiece;
    private int numPieces;
    private int piecesCount;
    private String dirPath;

    public FileManager(int peerId, CommonConfig config, boolean hasFile) {
        this.peerId = peerId;
        this.config = config;
        this.numPieces = config.getNumberOfPieces();
        this.pieces = new byte[numPieces][];
        this.hasPiece = new boolean[numPieces];
        this.piecesCount = 0;
        this.dirPath = "peer_" + peerId;

        new File(dirPath).mkdirs();

        if (hasFile) {
            loadFile();
        }
    }

    private void loadFile() {
        String filePath = dirPath + File.separator + config.fileName;
        try (RandomAccessFile raf = new RandomAccessFile(filePath, "r")) {
            for (int i = 0; i < numPieces; i++) {
                int size = getPieceSize(i);
                pieces[i] = new byte[size];
                raf.readFully(pieces[i]);
                hasPiece[i] = true;
                piecesCount++;
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public synchronized int getPieceSize(int index) {
        if (index == numPieces - 1) {
            int lastSize = config.fileSize % config.pieceSize;
            return lastSize == 0 ? config.pieceSize : lastSize;
        }
        return config.pieceSize;
    }

    public synchronized boolean hasPiece(int index) {
        return hasPiece[index];
    }

    public synchronized boolean hasAllPieces() {
        return piecesCount == numPieces;
    }

    public synchronized int getPiecesCount() {
        return piecesCount;
    }

    public synchronized byte[] getPiece(int index) {
        return pieces[index];
    }

    public synchronized void setPiece(int index, byte[] data) {
        if (!hasPiece[index]) {
            pieces[index] = data;
            hasPiece[index] = true;
            piecesCount++;

            if (hasAllPieces()) {
                saveFile();
            }
        }
    }

    public synchronized byte[] getBitfield() {
        int bitfieldSize = (int) Math.ceil((double) numPieces / 8);
        byte[] bitfield = new byte[bitfieldSize];
        for (int i = 0; i < numPieces; i++) {
            if (hasPiece[i]) {
                bitfield[i / 8] |= (1 << (7 - i % 8));
            }
        }
        return bitfield;
    }

    public int getNumPieces() {
        return numPieces;
    }

    private void saveFile() {
        String filePath = dirPath + File.separator + config.fileName;
        try (FileOutputStream fos = new FileOutputStream(filePath)) {
            for (int i = 0; i < numPieces; i++) {
                fos.write(pieces[i]);
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }
}
